stock_storage_type/models/stock_location.py

```diff
@@ -153,125 +251,109 @@ def get_storage_locations(self, products=None):
+        pertinent_loc_storage_types = LocStorageType.search(pertinent_loc_s_t_domain)
 
-    def _existing_quants(self, products=None, lot=None):
-        base_domain = [("location_id", "child_of", self.id)]
-        domain = self._prepare_existing_domain(base_domain, products=products, lot=lot)
-        return self.env["stock.quant"].search(domain, limit=1)
+        # now loop over the pertinent location storage types (there should be
+        # few of them) and check for properties to find suitable locations
+        valid_location_ids = set()
+        for loc_storage_type in pertinent_loc_storage_types:
+            location_domain = loc_storage_type._domain_location_storage_type(
+                compatible_locations, quants, products
+            )
+            _logger.debug("pertinent location domain: %s", location_domain)
+            locations = self.search(location_domain, limit=limit)
+            valid_location_ids |= set(locations.ids)
 
-    def _existing_planned_move_lines(self, products=None, lot=None):
-        base_domain = [
-            ("location_dest_id", "child_of", self.id),
-            ("state", "in", ("waiting", "partially_available", "assigned")),
-        ]
-        domain = self._prepare_existing_domain(base_domain, products=products, lot=lot)
-        return self.env["stock.move.line"].search(domain, limit=1)
+        valid_locations = self.env["stock.location"].search(
+            [("id", "in", list(valid_location_ids))], limit=limit
+        )
 
-    def _existing_planned_moves(self, products=None):
-        if self.child_ids:
-            # If a location is a leaf, it's a "bin", we know that the move line will
-            # be in this exact location. If it has sub-locations, we can't be sure
-            # where it will go and we don't want to restrict based on this.
-            return self.env["stock.move"].browse()
-        base_domain = [
-            ("location_dest_id", "=", self.id),
-            ("state", "in", ("waiting", "confirmed")),
-        ]
-        domain = self._prepare_existing_domain(base_domain, products=products)
-        return self.env["stock.move"].search(domain, limit=1)
+        _logger.debug(
+            "select allowed location for package storage"
+            " type %s (q=%s, p=%s) found %d locations",
+            package_storage_type.name,
+            quants,
+            products.mapped("name"),
+            len(valid_locations),
+        )
+        return valid_locations
 
-    def _prepare_existing_domain(self, base_domain, products=None, lot=None):
-        domain = base_domain
-        if products is not None:
-            extra_domain = [("product_id", "not in", products.ids)]
-            if lot is not None:
-                extra_domain = OR([extra_domain, [("lot_id", "!=", lot.id)]])
-            domain = AND([domain, extra_domain])
-        return domain
+    def _get_ordered_children_locations(self):
```

Comment:
I think we want only leaf locations and not the intermediate nodes in the location structure. Same for ABC

Reply:
Yes! That's exactly what I was discussing with @grindtildeath at the moment and started to change